### PR TITLE
Check the rsync as the first preparation step

### PR DESCRIPTION
**Pull request by psss**
As the very first prepare action for each guest we need to ensure
that rsync is installed on the guest if needed. This fixes failing
`/tests/run/worktree` when run against virtual provision method.

### CODE REVIEW

**Comment by psss** [comment]
This fix is necessary in order to make the new `worktree` feature correclty working with `virtual` provision. Please, review as soon as possible. Thanks.

**Review by jscotka — APPROVED** [review]
Seems well. just one nit pick. I think that it should not be solved anyhow by `prepare`. But theoretically these dependencies important for test execution, should be solved as part of `provision` step, what will install `rsync` or whatever as part of image/container init.

File: tmt/steps/prepare/__init__.py [hunk 1]
```diff
@@ -88,9 +93,14 @@ def go(self):
 
         # Prepare guests (including workdir sync)
         for guest in self.plan.provision.guests():
+            # Make sure rsync is installed, push the workdir
+            if rsync_required:
+                self.debug('Ensure that rsync is installed on the guest.')
+                guest.execute('rpm -q rsync || yum install -y rsync')
```
Comment (pvalena) [inline tmt/steps/prepare/__init__.py:99]: What about `dnf`?

**Review by pvalena — CHANGES_REQUESTED** [review]
You should use the method for installation we already have.

Otherwise LGTM.

File: tmt/steps/prepare/__init__.py (hunk 1 above)
Comment (psss) [reply tmt/steps/prepare/__init__.py:99]: I tested with Fedora images, `yum` seems to be always present and we need this working for older distros as well.

**Comment by psss** [comment]
> You should use the method for installation we already have.

Yes, I was pondering about this for some time. But this would mean that the extra `rsync` prepare plugin would be always included under the prepare step, not brining much benefit I believe. I finished with this approach as not causing unnecessary spam to the user.

File: tmt/steps/prepare/__init__.py (hunk 1 above)
Comment (jscotka) [reply tmt/steps/prepare/__init__.py:99]: `yum`  is also as backward compatibility there so `yum` should be good enough. I've also though about that point

**Review by FrNecas — APPROVED** [review]

**Review by hegerj — APPROVED** [review]
LGTM

**Comment by psss** [comment]
> But theoretically these dependencies important for test execution, should be solved as part of `provision` step, what will install `rsync` or whatever as part of image/container init.

The `prepare` step seems like a logical place for the initial guest preparation to me. Why do you think `provision` step would be better for this?

**Comment by jscotka** [comment]
> > But theoretically these dependencies important for test execution, should be solved as part of `provision` step, what will install `rsync` or whatever as part of image/container init.
> 
> The `prepare` step seems like a logical place for the initial guest preparation to me. Why do you think `provision` step would be better for this?

Because for me preparation is preparation for test execution, so there should be just items what are relevant to testplans, not some underlying tools, what are theoretically not anyhow important for system preparation steps. E.g. I can imagine to replace rsync, by using pure ssh like `tar cf - | ssh user@hostname tar xf -` and then rsync is unnecessary to install anyhow. So that each provision plugin should took care to be able to communicate with TMT. so just prepare plugin knows how, and what to install, so possible to be part of `push/pull` method as some check if system is prepared.

**Comment by psss** [comment]
> Because for me preparation is preparation for test execution, so there should be just items what are relevant to testplans, not some underlying tools, what are theoretically not anyhow important for system preparation steps.

Thanks for the detailed explanation. I see your point. We can polish this in the future. But let's not block the 1.4 on this now.

**Comment by jscotka** [comment]
> > Because for me preparation is preparation for test execution, so there should be just items what are relevant to testplans, not some underlying tools, what are theoretically not anyhow important for system preparation steps.
> 
> Thanks for the detailed explanation. I see your point. We can polish this in the future. But let's not block the 1.4 on this now.

Sure, it is generic RFE, not blocking this PR.

**Pull request merged**